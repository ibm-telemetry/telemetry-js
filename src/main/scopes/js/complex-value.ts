/*
 * Copyright IBM Corp. 2023, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

/**
 * Object representing a complex value.
 */
export class ComplexValue {
  constructor(public readonly complexValue: unknown) {}
}
