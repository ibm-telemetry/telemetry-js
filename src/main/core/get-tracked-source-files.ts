/*
 * Copyright IBM Corp. 2023, 2025
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { readFile } from 'node:fs/promises'
import * as path from 'node:path'

import * as ts from 'typescript'

import { type Logger } from './log/logger.js'
import { TrackedFileEnumerator } from './tracked-file-enumerator.js'
import { parseDocument } from 'htmlparser2'
import type { Document as HtmlDocument } from 'domhandler'
import { HtmlParsedFile } from '../scopes/wc/interfaces.js'

/**
 * Gets all tracked source files to consider for data collection,
 * filtered by supplied file extension array.
 *
 * @param cwd - Working directory for underlying command execution.
 * @param root - Root directory in which to search for tracked source files. This is an absolute
 * path.
 * @param logger - Logger instance to use.
 * @param fileExtensions - List of file extensions to filter files by.
 * @returns An array of source file objects.
 */
export async function getTrackedSourceFiles(
  cwd: string,
  root: string,
  logger: Logger,
  fileExtensions: string[]
) {
  logger.traceEnter('', 'getTrackedSourceFiles', [root, fileExtensions])

  const fileEnumerator = new TrackedFileEnumerator(logger)
  const files = []

  // If a file is passed instead of a directory, avoid the `git ls-tree` call
  if (fileExtensions.includes(path.extname(root))) {
    files.push(root)
  } else {
    files.push(
      ...(await fileEnumerator.find(cwd, root, (file) =>
        fileExtensions.includes(path.extname(file))
      ))
    )
  }

  const promises = files.map(async (file) => {
    return {
      fileName: file,
      async createSourceFile(): Promise<ts.SourceFile | HtmlDocument> {
        // Parse html file with `htmlparser2
        if (file.endsWith('html') || file.endsWith('htm')) {
          const parsedFile = parseDocument((await readFile(file)).toString()) as HtmlParsedFile
          parsedFile.fileName = file
        }

        // Parse any other files
        return ts.createSourceFile(
          file,
          (await readFile(file)).toString(),
          ts.ScriptTarget.ES2021,
          /* setParentNodes */ true
        )
      }
    }
  })

  const results = await Promise.all(promises)

  logger.traceExit(
    '',
    'getTrackedSourceFiles',
    results.map((result) => result.fileName)
  )
  return results
}
