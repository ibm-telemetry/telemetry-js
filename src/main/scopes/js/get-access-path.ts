/*
 * Copyright IBM Corp. 2024, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import * as ts from 'typescript'

import type { Logger } from '../../core/log/logger.js'
import { ComplexValue } from './complex-value.js'
import { getNodeValueHandler } from './node-value-handler-map.js'

/**
 * Constructs a JsToken object from a given Identifier type AST node.
 *
 * @param node - TS node to retrieve access path for.
 * @param sourceFile - Root AST node.
 * @param logger - A logger instance.
 * @param currAccessPath - For internal use only, tracks current constructed access path.
 * @param topLevel - For internal use only, tracks top level function call.
 * @returns Access path to node represented as string array (chunks).
 */
export default function getAccessPath(
  node: ts.Node,
  sourceFile: ts.SourceFile,
  logger: Logger,
  currAccessPath: string[] = [],
  topLevel: boolean = true
): string[] {
  switch (node.kind) {
    case ts.SyntaxKind.Identifier:
      return [...currAccessPath, (node as ts.Identifier).escapedText.toString()]
    case ts.SyntaxKind.PropertyAccessExpression:
      currAccessPath.push((node as ts.PropertyAccessExpression).name.escapedText.toString())
      break
    case ts.SyntaxKind.ElementAccessExpression: {
      const argumentExpression = (node as ts.ElementAccessExpression).argumentExpression
      const data = getNodeValueHandler(argumentExpression.kind, sourceFile, logger).getData(
        argumentExpression
      )
      if (data instanceof ComplexValue) {
        currAccessPath = [
          ...currAccessPath,
          getAccessPath(argumentExpression, sourceFile, logger, [], false).reverse().join('.')
        ]
      } else {
        currAccessPath.push(data?.toString() ?? '')
      }
      break
    }
  }

  let accessPath = currAccessPath

  if ('expression' in node) {
    accessPath = getAccessPath(
      node.expression as ts.Node,
      sourceFile,
      logger,
      currAccessPath,
      false
    )
  }

  return topLevel ? accessPath.reverse() : accessPath
}
