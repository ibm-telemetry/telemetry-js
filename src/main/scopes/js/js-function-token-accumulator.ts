/*
 * Copyright IBM Corp. 2023, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { JsFunction, JsImport, JsToken } from '../js/interfaces.js'
import { JsAccumulator } from '../js/js-accumulator.js'

/**
 * Responsible for maintaining an aggregated state of imports, functions and tokens.
 */
export class JsFunctionTokenAccumulator extends JsAccumulator {
  public readonly tokens: JsToken[]
  public readonly functions: JsFunction[]
  public readonly functionImports: Map<JsFunction, JsImport>
  public readonly tokenImports: Map<JsToken, JsImport>

  constructor() {
    super()
    this.tokens = []
    this.functions = []
    this.functionImports = new Map()
    this.tokenImports = new Map()
  }
}
