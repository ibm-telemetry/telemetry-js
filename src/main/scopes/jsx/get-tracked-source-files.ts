/*
 * Copyright IBM Corp. 2023, 2023
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import path from 'node:path'

import * as ts from 'typescript'

import { type Logger } from '../../core/log/logger.js'
import { TrackedFileEnumerator } from '../../core/tracked-file-enumerator.js'

/**
 * Gets all tracked source files to consider for data collection.
 *
 * @param root - Root directory in which to search for tracked source files.
 * @param logger - Logger instance to use.
 * @returns An array of source file objects.
 */
export async function getTrackedSourceFiles(root: string, logger: Logger) {
  const fileEnumerator = new TrackedFileEnumerator(logger)
  const allowedSuffixes = ['js', 'jsx', 'ts', 'tsx']

  const files = await fileEnumerator.find(root, (file) =>
    allowedSuffixes.includes(path.extname(file))
  )

  const program = ts.createProgram(files, {})

  return program.getSourceFiles().filter((file) => !file.isDeclarationFile)
}
