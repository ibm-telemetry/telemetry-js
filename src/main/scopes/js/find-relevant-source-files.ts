/*
 * Copyright IBM Corp. 2024, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import getPropertyByPath = require('lodash/get.js')
import * as path from 'node:path'

import { ObjectPath } from 'object-scan'

import { getTrackedSourceFiles } from '../../core/get-tracked-source-files.js'
import { Logger } from '../../core/log/logger.js'
import { NoInstallationFoundError } from '../../exceptions/no-installation-found-error.js'
import { getDependencyTree } from '../npm/get-dependency-tree.js'
import { getDirectoryPrefix } from '../npm/get-directory-prefix.js'
import { getInstalledVersionPaths } from '../npm/get-installed-version-paths.js'
import { getPackageData } from '../npm/get-package-data.js'
import { getPackageTrees } from '../npm/get-package-trees.js'
import { getTreePredecessor } from '../npm/get-tree-predecessor.js'
import { DependencyTree, PackageData } from '../npm/interfaces.js'

/**
 * Finds tracked source files and then filters them based on ones that appear in a  project which
 * depends on the in-context instrumented package/version.
 *
 * @param instrumentedPackage - Data about the instrumented package to use during filtering.
 * @param cwd - Current working directory. This must be inside of the root directory. This is an
 * absolute path.
 * @param root - Root-most directory. This is an absolute path.
 * @param fileExtensions - List of file extensions to capture metrics for.
 * @param logger - Logger instance.
 * @returns A (possibly empty) array of source files.
 */
export async function findRelevantSourceFiles(
  instrumentedPackage: PackageData,
  cwd: string,
  root: string,
  fileExtensions: string[],
  logger: Logger
) {
  logger.traceEnter('', 'findRelevantSourceFiles', [instrumentedPackage, cwd, root, fileExtensions])
  const sourceFiles = await getTrackedSourceFiles(root, logger, fileExtensions)

  const dependencyTree = await getDependencyTree(cwd, root, logger)

  const filterPromises = sourceFiles.map(async (f) => {
    const prefix = await getDirectoryPrefix(path.dirname(f.fileName), logger)
    const prefixPackageData = await getPackageData(prefix, root, logger)

    let packageTrees = getPackageTrees(dependencyTree, prefixPackageData)

    let instrumentedInstallVersions: string[] | undefined = undefined
    let shortestPathLength: number | undefined = undefined
    do {
      for (const tree of packageTrees) {
        const instrumentedInstallPaths = getInstalledVersionPaths(tree, instrumentedPackage.name)
        if (instrumentedInstallPaths.length > 0) {
          const pathsLength = instrumentedInstallPaths[0]?.length ?? 0
          if (shortestPathLength === undefined || pathsLength < shortestPathLength) {
            instrumentedInstallVersions = instrumentedInstallPaths.map(
              (path) => getPropertyByPath(tree, path)['version']
            )
            shortestPathLength = pathsLength
          }
        }
      }
      // did not find, go up one level for all packages
      packageTrees = packageTrees
        .map((tree) => getTreePredecessor(dependencyTree, tree['path'] as ObjectPath))
        .filter((tree) => tree !== undefined) as DependencyTree[]
    } while (shortestPathLength === undefined && packageTrees.length > 0)

    if (instrumentedInstallVersions === undefined) {
      throw new NoInstallationFoundError(instrumentedPackage.name)
    }

    return instrumentedInstallVersions.some((version) => version === instrumentedPackage.version)
  })
  const filterData = await Promise.all(filterPromises)

  const results = sourceFiles.filter((_, index) => {
    return filterData[index]
  })

  logger.traceEnter('', 'findRelevantSourceFiles', results)
  return results
}
