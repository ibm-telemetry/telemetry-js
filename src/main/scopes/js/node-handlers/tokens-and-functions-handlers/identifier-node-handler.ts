/*
 * Copyright IBM Corp. 2023, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import type * as ts from 'typescript'

import { JsToken } from '../../interfaces.js'
import { JsFunctionTokenAccumulator } from '../../js-function-token-accumulator.js'
import { JsNodeHandler } from '../js-node-handler.js'

/**
 * Holds logic to construct a JsToken object given an Identifier node.
 *
 */
export class IdentifierNodeHandler extends JsNodeHandler<JsToken> {
  /**
   * Processes a Identifier node data and adds it to the given accumulator.
   *
   * @param node - Node element to process.
   * @param accumulator - JsFunctionTokenAccumulator instance
   * that holds the aggregated tokens state.
   */
  handle(node: ts.Identifier, accumulator: JsFunctionTokenAccumulator) {
    accumulator.tokens.push(this.getData(node))
  }

  /**
   * Constructs a JsToken object from a given Identifier type AST node.
   *
   * @param node - Node element to process.
   * @returns Constructed JsToken object.
   */
  getData(node: ts.Identifier): JsToken {
    // TODO: implement, how to know this is not a part of a
    // CallExpression, PropertyAccessExpression or ElementAccessExpression?
    // TODO: test
    return {
      name: node.escapedText.toString(),
      accessPath: [node.escapedText.toString()],
      startPos: node.pos,
      endPos: node.end
    }
  }
}
