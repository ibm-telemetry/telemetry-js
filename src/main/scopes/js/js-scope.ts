/*
 * Copyright IBM Corp. 2024, 2025
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import type { ConfigSchema } from '@ibm/telemetry-config-schema'
import type * as ts from 'typescript'

import { Trace } from '../../core/log/trace.js'
import { Scope } from '../../core/scope.js'
import { EmptyScopeError } from '../../exceptions/empty-scope.error.js'
import { findRelevantSourceFiles } from '../js/find-relevant-source-files.js'
import type { JsFunction, JsImportMatcher, JsToken } from '../js/interfaces.js'
import { processFile } from '../js/process-file.js'
import { removeIrrelevantImports } from '../js/remove-irrelevant-imports.js'
import { getPackageData } from '../npm/get-package-data.js'
import type { PackageData } from '../npm/interfaces.js'
import { JsAllImportMatcher } from './import-matchers/js-all-import-matcher.js'
import { JsNamedImportMatcher } from './import-matchers/js-named-import-matcher.js'
import { JsRenamedImportMatcher } from './import-matchers/js-renamed-import-matcher.js'
import { JsFunctionTokenAccumulator } from './js-function-token-accumulator.js'
import { jsNodeHandlerMap } from './js-node-handler-map.js'
import { FunctionMetric } from './metrics/function-metric.js'
import { TokenMetric } from './metrics/token-metric.js'

/**
 * Scope class dedicated to data collection from a jsx environment.
 */
export class JsScope extends Scope {
  public override name = 'js' as const
  private runSync = false
  public static readonly fileExtensions = [
    '.jsx',
    '.mjs',
    '.js',
    '.cjs',
    '.tsx',
    '.mts',
    '.ts',
    '.cts',
    '.mjsx',
    '.cjsx',
    '.mtsx',
    '.ctsx'
  ]

  /**
   * Entry point for the scope. All scopes run asynchronously.
   */
  @Trace()
  public override async run(): Promise<void> {
    const collectorKeys = this.config.collect[this.name]
    if (collectorKeys === undefined || Object.keys(collectorKeys).length === 0) {
      throw new EmptyScopeError(this.name)
    }

    await this.captureAllMetrics(collectorKeys)
  }

  /**
   * Generates metrics for all discovered instrumented js functions and tokens
   * found in the current working directory's project, depending on config.
   *
   * @param collectorKeys - Keys for JS collection as defined by the current config.
   */
  @Trace()
  async captureAllMetrics(
    collectorKeys: NonNullable<ConfigSchema['collect']['js']>
  ): Promise<void> {
    const importMatchers: JsImportMatcher<JsToken | JsFunction>[] = [
      new JsAllImportMatcher(),
      new JsNamedImportMatcher(),
      new JsRenamedImportMatcher()
    ]

    const instrumentedPackage = await getPackageData(this.cwd, this.cwd, this.logger)

    const sourceFiles = await findRelevantSourceFiles(
      instrumentedPackage,
      this.cwd,
      this.root,
      JsScope.fileExtensions,
      this.logger
    )

    this.logger.debug('Filtered source files: ' + sourceFiles.map((f) => f.fileName))

    const promises: Promise<void>[] = []

    for (const sourceFile of sourceFiles) {
      const resultPromise = this.captureFileMetrics(
        (await sourceFile.createSourceFile()) as ts.SourceFile,
        instrumentedPackage,
        importMatchers,
        collectorKeys
      )

      if (this.runSync) {
        await resultPromise
      } else {
        promises.push(resultPromise)
      }
    }

    await Promise.allSettled(promises)
  }

  /**
   * Generates metrics for all discovered instrumented js functions and tokens found
   * in the supplied SourceFile node, depending on config.
   *
   * @param sourceFile - The sourcefile node to generate metrics for.
   * @param instrumentedPackage - Name and version of the instrumented package
   * to capture metrics for.
   * @param importMatchers - Matchers instances to use for import-function/token matching.
   * @param collectorKeys - Config keys defined for JS scope.
   */
  async captureFileMetrics(
    sourceFile: ts.SourceFile,
    instrumentedPackage: PackageData,
    importMatchers: JsImportMatcher<JsToken | JsFunction>[],
    collectorKeys: NonNullable<ConfigSchema['collect']['js']>
  ) {
    const accumulator = new JsFunctionTokenAccumulator()

    processFile(accumulator, sourceFile, jsNodeHandlerMap, this.logger)

    this.deduplicateFunctions(accumulator)

    this.deduplicateTokens(accumulator)

    removeIrrelevantImports(accumulator, instrumentedPackage.name)

    const promises: Array<Promise<void>> = []

    Object.keys(collectorKeys).forEach((key) => {
      switch (key) {
        case 'tokens':
          promises.push(
            this.captureTokenFileMetrics(accumulator, instrumentedPackage, importMatchers)
          )
          break
        case 'functions':
          promises.push(
            this.captureFunctionFileMetrics(accumulator, instrumentedPackage, importMatchers)
          )
          break
      }
    })

    await Promise.allSettled(promises)
  }

  /**
   * Generates metrics for all discovered instrumented js tokens found
   * and stored in the given accumulator.
   *
   * @param accumulator - Accumulator containing all previously computed imports and tokens.
   * @param instrumentedPackage - Name and version of the instrumented package
   * to capture metrics for.
   * @param importMatchers - Matchers instances to use for import-token matching.
   */
  async captureTokenFileMetrics(
    accumulator: JsFunctionTokenAccumulator,
    instrumentedPackage: PackageData,
    importMatchers: JsImportMatcher<JsToken>[]
  ) {
    this.resolveTokenImports(accumulator, importMatchers)

    accumulator.tokens.forEach((jsToken) => {
      const jsImport = accumulator.tokenImports.get(jsToken)

      if (jsImport === undefined) {
        return
      }

      this.capture(new TokenMetric(jsToken, jsImport, instrumentedPackage, this.logger))
    })
  }

  /**
   * Generates metrics for all discovered instrumented js functions found
   * and stored in the given accumulator.
   *
   * @param accumulator - Accumulator containing all previously computed imports and functions.
   * @param instrumentedPackage - Name and version of the instrumented package
   * to capture metrics for.
   * @param importMatchers - Matchers instances to use for import-function matching.
   */
  async captureFunctionFileMetrics(
    accumulator: JsFunctionTokenAccumulator,
    instrumentedPackage: PackageData,
    importMatchers: JsImportMatcher<JsFunction>[]
  ) {
    this.resolveFunctionImports(accumulator, importMatchers)

    accumulator.functions.forEach((jsFunction) => {
      const jsImport = accumulator.functionImports.get(jsFunction)

      if (jsImport === undefined) {
        return
      }

      this.capture(
        new FunctionMetric(jsFunction, jsImport, instrumentedPackage, this.config, this.logger)
      )
    })
  }

  resolveTokenImports(
    accumulator: JsFunctionTokenAccumulator,
    tokenMatchers: JsImportMatcher<JsToken>[]
  ) {
    accumulator.tokens.forEach((jsToken) => {
      const jsImport = tokenMatchers
        .map((tokenMatcher) => tokenMatcher.findMatch(jsToken, accumulator.imports))
        .find((jsImport) => jsImport !== undefined)

      if (jsImport === undefined) {
        return
      }

      accumulator.tokenImports.set(jsToken, jsImport)
    })
  }

  resolveFunctionImports(
    accumulator: JsFunctionTokenAccumulator,
    functionMatchers: JsImportMatcher<JsFunction>[]
  ) {
    accumulator.functions.forEach((jsFunction) => {
      const jsImport = functionMatchers
        .map((functionMatcher) => functionMatcher.findMatch(jsFunction, accumulator.imports))
        .find((jsImport) => jsImport !== undefined)

      if (jsImport === undefined) {
        return
      }

      accumulator.functionImports.set(jsFunction, jsImport)
    })
  }

  deduplicateFunctions(accumulator: JsFunctionTokenAccumulator) {
    accumulator.functions = accumulator.functions.filter(
      (f) =>
        !accumulator.functions.some(
          (func) => func.startPos >= f.startPos && func.endPos <= f.endPos && func !== f
        )
    )
  }

  deduplicateTokens(accumulator: JsFunctionTokenAccumulator) {
    // Given: foo.bar().baz
    // foo.bar().baz <- filtered
    // foo.bar <-- function, not touched
    accumulator.tokens = accumulator.tokens.filter(
      (token) =>
        !accumulator.functions.some(
          (func) => func.startPos >= token.startPos && func.endPos <= token.endPos
        )
    )
  }

  /**
   * **For testing purposes only.**
   * Makes the JsxScope collection run "synchronously" (one source file at a time). Defaults to
   * `false`.
   *
   * @param runSync - Boolean of whether or not to run synchronously.
   */
  setRunSync(runSync: boolean) {
    this.runSync = runSync
  }
}
