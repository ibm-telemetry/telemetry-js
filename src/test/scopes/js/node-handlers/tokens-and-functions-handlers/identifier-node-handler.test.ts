/*
 * Copyright IBM Corp. 2023, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import * as ts from 'typescript'
import { describe, expect, it } from 'vitest'

import { getTrackedSourceFiles } from '../../../../../main/core/get-tracked-source-files.js'
import { JsFunctionTokenAccumulator } from '../../../../../main/scopes/js/js-function-token-accumulator.js'
import { JsScope } from '../../../../../main/scopes/js/js-scope.js'
import { IdentifierNodeHandler } from '../../../../../main/scopes/js/node-handlers/tokens-and-functions-handlers/identifier-node-handler.js'
import { findNodesByType } from '../../../../__utils/find-nodes-by-type.js'
import { Fixture } from '../../../../__utils/fixture.js'
import { initLogger } from '../../../../__utils/init-logger.js'

describe('class: IdentifierNodeHandler', async () => {
  const logger = initLogger()
  const fixture = new Fixture('js-samples/all-js-token-types.ts')
  const sourceFile = (
    await getTrackedSourceFiles(fixture.path, logger, JsScope.fileExtensions)
  )[0] as ts.SourceFile

  const handler = new IdentifierNodeHandler(sourceFile, logger)

  it('correctly returns the JsTokens for a complex fixture', async () => {
    const accumulator = new JsFunctionTokenAccumulator()
    const identifiers = findNodesByType(sourceFile, ts.SyntaxKind.Identifier)

    identifiers.forEach((identifier) => {
      handler.handle(identifier as ts.Identifier, accumulator)
    })

    expect(accumulator.tokens).toMatchSnapshot()
  })
})
