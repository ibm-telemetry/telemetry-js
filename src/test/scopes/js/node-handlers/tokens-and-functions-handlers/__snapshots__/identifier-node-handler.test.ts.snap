// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`class: IdentifierNodeHandler > correctly returns the JsTokens for a complex fixture 1`] = `
[
  {
    "accessPath": [
      "blue",
    ],
    "endPos": 197,
    "name": "blue",
    "startPos": 0,
  },
  {
    "accessPath": [
      "blue",
    ],
    "endPos": 211,
    "name": "blue",
    "startPos": 203,
  },
  {
    "accessPath": [
      "colors",
    ],
    "endPos": 229,
    "name": "colors",
    "startPos": 219,
  },
  {
    "accessPath": [
      "blue",
    ],
    "endPos": 234,
    "name": "blue",
    "startPos": 230,
  },
  {
    "accessPath": [
      "colors",
    ],
    "endPos": 242,
    "name": "colors",
    "startPos": 234,
  },
  {
    "accessPath": [
      "blue",
    ],
    "endPos": 247,
    "name": "blue",
    "startPos": 243,
  },
  {
    "accessPath": [
      "blue",
    ],
    "endPos": 267,
    "name": "blue",
    "startPos": 259,
  },
  {
    "accessPath": [
      "gradient",
    ],
    "endPos": 283,
    "name": "gradient",
    "startPos": 275,
  },
  {
    "accessPath": [
      "big",
    ],
    "endPos": 288,
    "name": "big",
    "startPos": 285,
  },
  {
    "accessPath": [
      "blue",
    ],
    "endPos": 296,
    "name": "blue",
    "startPos": 288,
  },
  {
    "accessPath": [
      "THE_CONSTANT",
    ],
    "endPos": 315,
    "name": "THE_CONSTANT",
    "startPos": 303,
  },
  {
    "accessPath": [
      "bla",
    ],
    "endPos": 340,
    "name": "bla",
    "startPos": 336,
  },
  {
    "accessPath": [
      "THE_CONSTANT",
    ],
    "endPos": 355,
    "name": "THE_CONSTANT",
    "startPos": 342,
  },
  {
    "accessPath": [
      "anObject",
    ],
    "endPos": 390,
    "name": "anObject",
    "startPos": 355,
  },
  {
    "accessPath": [
      "BLA",
    ],
    "endPos": 394,
    "name": "BLA",
    "startPos": 391,
  },
]
`;
